code/game/objects/machinery/vending/vending.dm

```diff
@@ -610,63 +626,159 @@
+			if((A.flags_attach_features & ATTACH_REMOVABLE) && !(is_type_in_list(A, G.starting_attachment_types))) //There are attachments that are default and others that can't be removed
+				display_message_and_visuals(user, show_feedback, "Gun has non-standard attachments equipped!", VENDING_RESTOCK_DENY)
+				return FALSE
+
+	else if(record.amount >= 0) //Item is finite so we are more strict on it's condition
+
+		if(isammomagazine(item_to_stock))
 			var/obj/item/ammo_magazine/A = item_to_stock
 			if(A.current_rounds < A.max_rounds)
-				to_chat(user, span_warning("[A] isn't full. Fill it before you can restock it."))
-				return
-		else if(istype(item_to_stock, /obj/item/cell))
+				display_message_and_visuals(user, show_feedback, "Magazine isn't full!", VENDING_RESTOCK_DENY)
+				return FALSE
+
+		if(iscell(item_to_stock))
 			var/obj/item/cell/cell = item_to_stock
+
 			if(cell.charge < cell.maxcharge)
-				to_chat(user, span_warning("\The [cell] isn't full. You must recharge it before you can restock it."))
-				return
-		else if(isitemstack(item_to_stock))
+
+				if(vending_flags & VENDING_RECHARGER) // Item is finite and not full. Time to try to recharge
+					recharge_amount = cell.maxcharge - cell.charge
+
+					if(machine_current_charge == 0)
+						display_message_and_visuals(user, show_feedback, "No power!", VENDING_RESTOCK_DENY)
+						return FALSE
+
+					else if(machine_current_charge < recharge_amount) // Not enough but some charge remaining so partially recharge cell and move on
+						cell.give(machine_current_charge)
+						machine_current_charge = 0
+						cell.update_icon()
+						display_message_and_visuals(user, show_feedback, "Cell charged partially! [round(cell.percent())]%.", VENDING_RESTOCK_RECHARGE)
+						playsound(loc, 'sound/machines/hydraulics_1.ogg', 25, 0, 1)
+						return FALSE
+
+				else
+					display_message_and_visuals(user, show_feedback, "Cell isn't at full charge!", VENDING_RESTOCK_DENY)
+					return FALSE
+
+		if(isitemstack(item_to_stock))
 			var/obj/item/stack/stack = item_to_stock
 			if(stack.amount != initial(stack.amount))
-				to_chat(user, span_warning("[stack] has been partially used. You must replace the missing amount before you can restock it."))
-				return
-		else if(istype(item_to_stock, /obj/item/reagent_containers))
+				display_message_and_visuals(user, show_feedback, "[stack] has been partially used. Refill it!", VENDING_RESTOCK_DENY)
+				return FALSE
+
+		if(isreagentcontainer(item_to_stock))
 			var/obj/item/reagent_containers/reagent_container = item_to_stock
 			if(!reagent_container.free_refills && !reagent_container.has_initial_reagents())
-				to_chat(user, span_warning("\The [reagent_container] is missing some of its reagents. You must refill it before you can restock it."))
-				return
+				display_message_and_visuals(user, show_feedback, "\The [reagent_container] is missing some of its reagents!", VENDING_RESTOCK_DENY)
+				return FALSE
+
+	if(isgrenade(item_to_stock))
```

Comment:
else if here

Reply:
Moved it up right below the isgun check and made it else if. Making it else if there would conflict with the finite item check and allow the grenades to be restocked even when primed.

Reply:
Put the grenade check under the isgun check as an elseif. Having it be elseif right there where it was would conflict with the finite item check and allow people to restock grenades even when primed.